crypto/txsigner_eip155.go

```diff
@@ -2,90 +2,166 @@ package crypto
+
+	// RLP(nonce, gasPrice, -, -, -, -, -, -, -)
+	hashPreimage.Set(RLP.NewBigInt(tx.GasPrice()))
+
+	// RLP(nonce, gasPrice, gas, -, -, -, -, -, -)
+	hashPreimage.Set(RLP.NewUint(tx.Gas()))
+
+	// Checking whether the transaction is a smart contract deployment
+	if tx.To() == nil {
+		// RLP(nonce, gasPrice, gas, to, -, -, -, -, -)
+		hashPreimage.Set(RLP.NewNull())
+	} else {
+		// RLP(nonce, gasPrice, gas, to, -, -, -, -, -)
+		hashPreimage.Set(RLP.NewCopyBytes((*(tx.To())).Bytes()))
 	}
 
-	if vv := bigV.Uint64(); bits.Len(uint(vv)) <= 8 {
-		protected = vv != 27 && vv != 28
+	// RLP(nonce, gasPrice, gas, to, value, -, -, -, -)
+	hashPreimage.Set(RLP.NewBigInt(tx.Value()))
+
+	// RLP(nonce, gasPrice, gas, to, value, input, -, -, -)
+	hashPreimage.Set(RLP.NewCopyBytes(tx.Input()))
+
+	// RLP(nonce, gasPrice, gas, to, value, input, chainId, -, -)
+	hashPreimage.Set(RLP.NewUint(signer.chainID))
+
+	// RLP(nonce, gasPrice, gas, to, value, input, chainId, 0, -)
+	hashPreimage.Set(RLP.NewUint(0))
+
+	// RLP(nonce, gasPrice, gas, to, value, input, chainId, 0, 0)
+	hashPreimage.Set(RLP.NewUint(0))
+
+	// keccak256(RLP(nonce, gasPrice, gas, to, value, input))
+	hash = keccak.Keccak256Rlp(nil, hashPreimage)
+
+	arenaPool.Put(RLP)
+
+	return types.BytesToHash(hash)
+}
+
+// Sender returns the sender of the transaction
+func (signer *EIP155Signer) Sender(tx *types.Transaction) (types.Address, error) {
+	if tx.ChainID() == nil || tx.ChainID().Cmp(big.NewInt(0)) == 0 {
+		return signer.HomesteadSigner.Sender(tx)
 	}
 
-	if !protected {
-		return (&FrontierSigner{}).Sender(tx)
+	v, r, s := tx.RawSignatureValues()
+
+	// Checking one of the values is enought since they are inseparable
+	if v == nil {
+		return types.Address{}, errors.New("Sender method: Unknown signature")
 	}
 
-	// Reverse the V calculation to find the original V in the range [0, 1]
-	// v = CHAIN_ID * 2 + 35 + {0, 1}
-	mulOperand := big.NewInt(0).Mul(big.NewInt(int64(e.chainID)), big.NewInt(2))
-	bigV.Sub(bigV, mulOperand)
-	bigV.Sub(bigV, big35)
+	// Reverse the V calculation to find the parity of the Y coordinate
+	// v = CHAIN_ID * 2 + 35 + {0, 1} -> {0, 1} = v - 35 - CHAIN_ID * 2
+
+	a := big.NewInt(0)
+	b := big.NewInt(0)
+	parity := big.NewInt(0)
+
+	// a = v - 35
+	a.Sub(v, big35)
+
+	// b = CHAIN_ID * 2
+	b.Mul(big.NewInt(int64(signer.chainID)), big.NewInt(2))
 
-	return recoverAddress(e.Hash(tx), r, s, bigV, e.isHomestead)
+	// parity = a - b
+	parity.Sub(a, b)
+
+	return recoverAddress(signer.Hash(tx), r, s, parity, true)
 }
 
-// SignTx signs the transaction using the passed in private key
-func (e *EIP155Signer) SignTx(
-	tx *types.Transaction,
-	privateKey *ecdsa.PrivateKey,
-) (*types.Transaction, error) {
+// SingTx takes the original transaction as input and returns its signed version
+func (signer *EIP155Signer) SignTx(tx *types.Transaction, privateKey *ecdsa.PrivateKey) (*types.Transaction, error) {
+	if tx.ChainID() == nil || tx.ChainID().Cmp(big.NewInt(0)) == 0 {
```

Comment:
Like I said, this condition is not correct in `Sender` and `SignTx` function. Remove this.